apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
 
-    return Promise.resolve();
+  var doSlideAnimation = oldView && newView &&
+    oldPanelElement !== newPanelElement;
+  var isGoingBack = oldView && newView && oldView.previous === newView.name;
+
+  if (doSlideAnimation) {
+    newPanelElement.classList.add('panel-will-activate', 'panel-active');
+    oldPanelElement.classList.add('panel-will-deactivate');
+
+    newPanelElement.classList.toggle('panel-animation-back', isGoingBack);
+    oldPanelElement.classList.toggle('panel-animation-back', isGoingBack);
+
+    return waitForSlideAnimation(newPanelElement).then(() => {
+      oldPanelElement.classList.remove(
+        'panel-will-activate', 'panel-will-deactivate',
+        'panel-active', 'panel-animation-back'
+      );
+      newPanelElement.classList.remove(
+        'panel-will-deactivate', 'panel-will-activate', 'panel-animation-back'
+      );
+    }, () => {
+      oldPanelElement.classList.remove(
+        'panel-will-deactivate', 'panel-animation-back'
+      );
+      newPanelElement.classList.remove(
+        'panel-will-activate', 'panel-animation-back'
+      );
+    });
+  }
+
+  if (oldView) {
+    oldPanelElement.classList.remove('panel-active');
+  }
+
+  if (newView) {
+    newPanelElement.classList.add('panel-active');
+  }
+
+  return Promise.resolve();
+}
+
+/**
+ * Reacts to hash changes.
+ * 
+ * This is the heart of the navigation logic and handles running the various
+ * naviation steps.
+ * It also handles any unwanted errors. (TODO: needs to tune this a little
+ * more).
+ *
+ * @param {Event} e The hashchange event (not actually used in the function)
+ */
+function onHashChange(e) {
+  debug('onHashChange, location=', window.location);
+  var newPanel = findViewFromLocation(window.location);
+  if (!newPanel) {
+    return;
+  }
+
+  var args = Utils.params(decodeURIComponent(window.location.hash));
+
+  startNavigation({ newPanel, args }).then(
+    () => executeNavigationStep('beforeLeave')
+  ).then(
+    () => executeNavigationStep('beforeEnter').catch(catchStepError)
+  ).then(
+    switchPanel
+  ).then(
+    setFutureState
+  ).then(
+    () => executeNavigationStep('afterEnter').catch(catchStepError)
+  ).then(
+    () => executeNavigationStep('afterLeave').catch(catchStepError)
+  ).then(
+    endNavigation
+  ).catch(
+    onNavigationError // TODO reset to previous hash or call back ?
+  );
+}
+
+/**
+ * Attaches the hashchange event handler.
+ */
+function attachHistoryListener() {
+  window.addEventListener('hashchange', onHashChange);
+}
+
+/**
+ * Removes the hashchange event handler.
+ */
+function detachHistoryListener() {
+  window.removeEventListener('hashchange', onHashChange);
+}
+
+exports.Navigation = {
+  /**
+   * Calls `window.back` after running the `beforeLeave` step.
+   *
+   * @returns {Promise} Resolves when the action is done.
+   */
+  back() {
+    debug('back()');
+    startNavigation({ partial: true }).then(
+      () => executeNavigationStep('beforeLeave')
```

Comment:
we need to call beforeLeave before window.back because we might change the document. And we don't want to change the document too soon :)